pkg/search/proxy/store/util.go

```diff
@@ -195,9 +197,67 @@ func (w *watchMux) AddSource(watcher watch.Interface, decorator func(watch.Event
 
 // Start run the watcher
 func (w *watchMux) Start() {
-	for _, source := range w.sources {
-		go w.startWatchSource(source.watcher, source.decorator)
+	// Build select cases dynamically. The final result is like:
+	//   select {
+	//   case v, ok := <- w.sources[0].watcher.ResultChan():
+	//	   ...
+	//   case v, ok := <- w.sources[1].watcher.ResultChan():
+	//	   ...
+	//   case v, ok := <- w.sources[2].watcher.ResultChan():
+	//	   ...
+	//   ...
+	//   case <- w.done:
+	//	   ...
+	//   }
+	cases := make([]reflect.SelectCase, len(w.sources)+1)
+	for i, source := range w.sources {
+		cases[i] = reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(source.watcher.ResultChan()),
+		}
+	}
+
+	// add done chan select case at last
+	doneCaseIndex := len(cases) - 1
+	cases[doneCaseIndex] = reflect.SelectCase{
+		Dir:  reflect.SelectRecv,
+		Chan: reflect.ValueOf(w.done),
 	}
+
+	go func() {
+		defer func() {
+			for _, source := range w.sources {
+				source.watcher.Stop()
+			}
+		}()
+		defer close(w.result)
+		for {
+			chosen, val, ok := reflect.Select(cases)
```

Comment:
The performance of `reflect.Select` is worse than that if `select`. if the scale of the federate cluster is too large, there may be performance issue.